eth_keys/libsecp256k1.nim

```diff
@@ -0,0 +1,305 @@
+    context: ptr secp256k1_context
+    error: string
+
+var ekContext {.threadvar.}: EthKeysContext
+  ## Thread local variable which holds current context
+
+##
+## Private procedures interface
+##
+
+proc illegalCallback(message: cstring; data: pointer) {.cdecl.} =
+  let ctx = cast[EthKeysContext](data)
+  ctx.error = $message
+
+proc errorCallback(message: cstring, data: pointer) {.cdecl.} =
+  let ctx = cast[EthKeysContext](data)
+  ctx.error = $message
+
+proc newEthKeysContext(): EthKeysContext =
+  ## Create new `EthKeysContext`.
+  result = new EthKeysContext
+  let flags = cuint(SECP256K1_CONTEXT_VERIFY or SECP256K1_CONTEXT_SIGN)
+  result.context = secp256k1_context_create(flags)
+  secp256k1_context_set_illegal_callback(result.context, illegalCallback,
+                                         cast[pointer](result))
+  secp256k1_context_set_error_callback(result.context, errorCallback,
+                                       cast[pointer](result))
+  result.error = ""
+
+proc getSecpContext(): ptr secp256k1_context =
+  ## Get current `secp256k1_context`
+  if isNil(ekContext):
+    ekContext = newEthKeysContext()
+  result = ekContext.context
+
+proc getContext(): EthKeysContext =
+  ## Get current `EccContext`
+  if isNil(ekContext):
+    ekContext = newEthKeysContext()
+  result = ekContext
+
+template raiseSecp256k1Error() =
+  ## Raises `libsecp256k1` error as exception
+  let mctx = getContext()
+  if len(mctx.error) > 0:
+    var msg = mctx.error
+    mctx.error.setLen(0)
+    raise newException(Secp256k1Exception, msg)
+
+proc libsecp256k1ErrorMsg(): string =
+  let mctx = getContext()
+  result = mctx.error
+
+proc setErrorMsg(m: string) =
+  let mctx = getContext()
+  mctx.error = m
+
+##
+## Public procedures interface
+##
+
+proc newPrivateKey*(): PrivateKey =
+  ## Generates new private key.
+  let ctx = getSecpContext()
+  while true:
+    if randomBytes(result.data) == KeyLength:
+      if secp256k1_ec_seckey_verify(ctx, cast[ptr cuchar](addr result)) == 1:
+        break
+
+proc getPublicKey*(seckey: PrivateKey): PublicKey =
+  ## Return public key for private key `seckey`.
+  let ctx = getSecpContext()
+  if secp256k1_ec_pubkey_create(ctx, addr result,
+                                cast[ptr cuchar](unsafeAddr seckey)) != 1:
+    raiseSecp256k1Error()
+
+proc safeGetPublicKey*(seckey: PrivateKey,
+                       pubkey: var PublicKey): EthKeysStatus =
+  ## Return public key associated with specified privat key `seckey`.
+  let ctx = getSecpContext()
+  if secp256k1_ec_pubkey_create(ctx, addr pubkey,
+                                cast[ptr cuchar](unsafeAddr seckey)) != 1:
+    result = EthKeysStatus.Error
+  else:
+    result = EthKeysStatus.Success
+
+proc newKeyPair*(): KeyPair =
+  ## Generates new private and public key.
+  result.seckey = newPrivateKey()
+  result.pubkey = result.seckey.getPublicKey()
+
+proc safeNewKeyPair*(keypair: var KeyPair): EthKeysStatus =
```

Comment:
Ditto, `newPrivateKey` is guaranteed to return a valid `PrivateKey`, but then it is passed to `safeGetPublicKey` in a pessimistic manner.

Reply:
As a matter of fact, I'm not sure if `KeyPair` type is useful at all, along with its creating functions. `PrivateKey` is already `KeyPair` enough, since you can always get a pub key from it if you need to. So that's just redundant, imo.